torch/csrc/distributed/rpc/request_callback.cpp

```diff
@@ -17,27 +17,38 @@ namespace {
 struct ClearAutogradContextGuard {
   ClearAutogradContextGuard() = default;
   ~ClearAutogradContextGuard() {
-    clear();
+    if (valid) {
+      DistAutogradContainer::getInstance().clearCurrentContext();
+    }
   }
 
-  void clear() {
-    auto& autogradContainer = DistAutogradContainer::getInstance();
-    autogradContainer.clearCurrentContext();
+  void cancel() {
+    valid = false;
   }
+  bool valid = true;
 };
 
 } // anonymous namespace
 
-Message RequestCallback::operator()(Message& request) const {
+std::shared_ptr<FutureMessage> RequestCallback::operator()(
+    Message& request) const {
   // For a recv thread, current context id should be invalid outside
   // processMessage().
-  ClearAutogradContextGuard guard;
   try {
-    return processMessage(request);
+    // do we need to run clear() in the future invocation?
+    ClearAutogradContextGuard guard;
```

Comment:
Same comments regarding guards and cancelling hold here.

Reply:
Right, reverted this to the original.